content/blog/2024/12/zephyr-how-to-use-psram/index.md

```diff
@@ -0,0 +1,587 @@
+
+***ESP_SPIRAM_MEMTEST:*** This configuration parameter controls SPIRAM memory test during initialization. It is enabled by default and can be disabled for faster startup.
+
+***SPIRAM_MODE:*** This configuration parameter selects the mode of SPI RAM chip in use. The permitted values are `SPIRAM_MODE_QUAD` and `SPIRAM_MODE_OCT`. Please note that SPIRAM_MODE_OCT is only available in ESP32-S3.
+
+***SPIRAM_TYPE:*** This configuration parameter defines the type of SPIRAM chip in use:
+
+| SPIRAM type | Your PSRAM capacity |
+| --- | --- |
+| `SPIRAM_TYPE_ESPPSRAM16` | 2 MB |
+| `SPIRAM_TYPE_ESPPSRAM32` | 4 MB |
+| `SPIRAM_TYPE_ESPPSRAM64` | 8 MB |
+
+***SPIRAM_SPEED:*** This configuration parameter sets the SPIRAM clock speed in MHz:
+ | Value | Clock speed |
+ | --- | --- |
+ | SPIRAM_SPEED_20M  |  20 MHz  |
+ | SPIRAM_SPEED_26M  |  26 MHz  |
+ | SPIRAM_SPEED_40M  |  40 MHz  |
+ | SPIRAM_SPEED_80M  |  80 MHz  |
+ | SPIRAM_SPEED_120M | 120 MHz  |
+
+***SPIRAM_FETCH_INSTRUCTIONS:*** This configuration parameter allows moving Instructions from flash to PSRAM. If enabled, instructions in flash will be moved into PSRAM on startup. If ```SPIRAM_RODATA``` parameter is also enabled, the code that normally requires execution during the SPI1 flash operation does not need to be placed in IRAM, thus optimizing RAM usage. By default, this parameter is disabled.
+
+***SPIRAM_RODATA:*** This configuration parameter allows moving read-only data from flash to PSRAM. If ```SPIRAM_FETCH_INSTRUCTIONS``` parameter is also enabled, the code that normally requires execution during the SPI1 flash operation  does not need to be placed in IRAM, thus optimizing RAM usage.
+
+## Installing Zephyr and its dependencies
+
+To install Zephyr RTOS and the necessary tools, follow the instructions in the Zephyr [Getting Started Guide](https://docs.zephyrproject.org/latest/develop/getting_started/index.html). By the end of the process, you will have a command-line Zephyr development environment set up and ready to build your application with `west` — the meta-tool responsible for building your application and flashing the generated binary, and other tasks.
+
+Additionally, you need to execute the following command to prepare your environment for building applications for Espressif SoCs:
+
+```sh
+west blobs fetch hal_espressif
+```
+
+## Dynamically Allocating PSRAM Memory
+
+PSRAM memory blocks can be made available to applications through Zephyr's shared multi-heap library.  The shared multi-heap memory pool manager uses the multi-heap allocator to manage a set of reserved memory regions with varying capabilities and attributes. For PSRAM, enabling the ```ESP_SPIRAM``` and ```SHARED_MULTI_HEAP``` parameters causes the external PSRAM to be mapped into the data virtual memory space during Zephyr's early initialization stage. The shared multi-heap framework is initialized, and the PSRAM memory region is added to the pool.
```

Comment:
library.  The (double spacing)
maybe use the editor tool for a search, review and replace is an easy fix
there are more cases below